compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs

```diff
@@ -438,33 +439,26 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
     pub(in super::super) fn suggest_boxing_when_appropriate(
         &self,
         err: &mut Diagnostic,
-        expr: &hir::Expr<'_>,
+        span: Span,
+        hir_id: HirId,
         expected: Ty<'tcx>,
         found: Ty<'tcx>,
     ) -> bool {
-        if self.tcx.hir().is_inside_const_context(expr.hir_id) {
-            // Do not suggest `Box::new` in const context.
+        let map = self.tcx.hir();
+        // Do not suggest `Box::new` in const context.
+        if map.is_inside_const_context(hir_id) || !expected.is_box() || found.is_box() {
             return false;
         }
-        if !expected.is_box() || found.is_box() {
-            return false;
-        }
-        let boxed_found = self.tcx.mk_box(found);
-        if self.can_coerce(boxed_found, expected) {
-            err.multipart_suggestion(
-                "store this in the heap by calling `Box::new`",
-                vec![
-                    (expr.span.shrink_to_lo(), "Box::new(".to_string()),
-                    (expr.span.shrink_to_hi(), ")".to_string()),
-                ],
-                Applicability::MachineApplicable,
-            );
-            err.note(
-                "for more on the distinction between the stack and the heap, read \
-                 https://doc.rust-lang.org/book/ch15-01-box.html, \
-                 https://doc.rust-lang.org/rust-by-example/std/box.html, and \
-                 https://doc.rust-lang.org/std/boxed/index.html",
-            );
+        if self.can_coerce(self.tcx.mk_box(found), expected) {
+            let suggest_boxing = match found.kind() {
+                ty::Tuple(tuple) if tuple.is_empty() => SuggestBoxing::Unit { start: span.shrink_to_lo(), end: span, },
+                ty::Generator(def_id, ..) if let Some(def_id) = def_id.as_local() && let Body {
+                    generator_kind: Some(GeneratorKind::Async(AsyncGeneratorKind::Closure)), ..
+                } = map.body(map.body_owned_by(def_id)) => SuggestBoxing::AsyncBody,
+                _ => SuggestBoxing::Other { start: span.shrink_to_lo(), end: span.shrink_to_hi() },
+            };
```

Comment:
Sorry, this is really hard to read without formatting :sweat_smile: 

```suggestion
            let suggest_boxing = match found.kind() {
                ty::Tuple(tuple) if tuple.is_empty() => SuggestBoxing::Unit {
                    start: span.shrink_to_lo(),
                    end: span,
                },
                ty::Generator(def_id, ..)
                    if let Some(def_id) = def_id.as_local()
                        && let Body {
                            generator_kind: Some(GeneratorKind::Async(AsyncGeneratorKind::Closure)),
                            ..
                        } = map.body(map.body_owned_by(def_id)) =>
                {
                    SuggestBoxing::AsyncBody
                }
                _ => SuggestBoxing::Other {
                    start: span.shrink_to_lo(),
                    end: span.shrink_to_hi(),
                },
            };
```

Reply:
Also, you should probably just use `tcx.generator_kind` here and a `match!`

Reply:
Thank you! i wish i had known about that earlier.